### ISSUE TITLE
[Question] Relaxed ordering in pop operation of lock free stack

### ISSUE THREAD

**Issue by dubuduru** [kaist-cp/cs431#761]
Hi, it might be too late to ask more question.. but your answer would help me a lot!

According to the slides, the ordering of cas in the pop operation ([L74](https://github.com/kaist-cp/cs431/blob/main/src/lockfree/stack.rs#L74)) can be relaxed because view of "next" is already released.
I understood the view of next is released when we push that node, and the released view is attached to self.head.
Also, what I understood is when we store a value on a location with Relaxed ordering, there should be no view attached to the location.

Suppose that we execute 2 pop operation.
At first pop, in L74, we atomically store "next" to self.head (suppose cas succeeded). Then what should be the view attached to self.head? As we just executed cas on self.head with Relaxed, would there be no view attached to it? Or would there be some released view which is pointwise-maximum of previous released view and the result of cas?

If there is no view attached to self.head, then second pop will load the value self.head with Acquire ordering, but no view will acquired. If so, we should use Release ordering when we cas in pop.

In short, my question is:
What should be the view attached to self.head after we cas on self.head with Relaxed ordering in [L74](https://github.com/kaist-cp/cs431/blob/main/src/lockfree/stack.rs#L74)?

Thank you!

**Comment by Lee-Janggun** [comment]
The view should be that of the thread that pushed the node to head. Since the view of that thread included the messages of all the nodes in the stack at that time, it is safe to access the rest of the stack with this view.

**Comment by dubuduru** [comment]
Thank you for your answer!
Can I ask why is that? I thought only Release attaches view to a message. Is it wrong and can Relaxed attach per-thread view to a message?

**Comment by Lee-Janggun** [comment]
It does not. Views are attached to values. Hence when the head is updated, the view is updated as well.

**Comment by dubuduru** [comment]
Now I understand the point I missed. Thank you :)
